### PR TITLE
Review of the rewrite

### PR DESCRIPTION
**Pull request by fehlmach**
- Removes inconsistencies
- A Zone can only contain Singular Assertion
- Remove UDP support for RAINS
- Simplifies some large sentences
- Fixes Types

### CODE REVIEW

File: draft-trammell-rains-protocol.md [hunk 1]
```diff
@@ -122,25 +122,25 @@ Its architecture ({{architecture}}) and information model
...
 - All assertions have validity within a specific context. A context determines
   the rules for chaining signatures to verify validity of an assertion. The
-  global context is a special case of context, which uses chains from the
-  global naming root key. The use of context explicitly separates global usage
-  of the DNS from local usage thereof, and allows other application-specific
-  naming constraints to be bound to names; see {{assertion-context}}.
-  Queries are valid in one or more contexts, with specific rules for
-  determining which assertions answer which queries; see
+  global context is a special case of context, which uses chains from one or
+  several global naming root key(s). The use of context explicitly separates
```
Comment (britram) [inline draft-trammell-rains-protocol.md]: As far as RAINS itself is concerned (in this document) there's one global root; the split root belongs in the rains-scion doc. Will fix after merge.

File: draft-trammell-rains-protocol.md [hunk 2]
```diff
@@ -170,14 +170,15 @@ In addition, the following terms are used in this document as defined:
 
 - Subject: A name or address about which Assertions can be made.
 - Object: A type/value pair of information about a name within an Assertion.
-- Assertion: A mapping between a Subject and an Object, signed by the Authority
-  for the namespace containing that Subject. See {{assertions}}.
+- Singular Assertion: A mapping between a Subject and one or several Objects,
```
Comment (britram) [inline draft-trammell-rains-protocol.md]: We still need a definition of "Assertion" as a superclass of these.

File: draft-trammell-rains-protocol.md [hunk 3]
```diff
@@ -291,9 +292,8 @@ are easily distinguishable from each other by its human users.
 
 A naming service should impose as little structure on the names it supports as
 practical in order to be universally applicable. Naming services that impose a
-given organizational structure on the names expressible using the service will
-not translate well to societies where that organizational structure is not
-prevalent.
+given organizational structure on the names will not translate well to societies
```
Comment (britram) [inline draft-trammell-rains-protocol.md]:
```suggestion
given organizational structure on names will not translate well to societies
```

File: draft-trammell-rains-protocol.md [hunk 4]
```diff
@@ -430,20 +428,19 @@ An Assertion which is not intended to be explicitly inconsistent by the
...
 
 ### Availability
 
-The naming service as a whole is resilient to failures of individual
-nodes providing the naming service, as well as to failures of links among
-them. Intentional prevention of successful, authenticated query by an
-adversary should be as hard as practical.
+The naming service as a whole is resilient to failures of individual nodes
+providing the naming service, as well as to failures of links among them.
+Intentional prevention of a successful answer to a query by an adversary should
```
Comment (britram) [inline draft-trammell-rains-protocol.md]:
```suggestion
Intentional prevention by an adversary of a successful answer to a query should
```

File: draft-trammell-rains-protocol.md [hunk 5]
```diff
@@ -605,62 +602,66 @@ capabilities-based versioning of the protocol, and for recognition of a chunk of
...
 rendered using alternate structured-data representations that support the
 concepts of maps and arrays. For example, YAML or JSON could be used to
 represent RAINS messages and data structures for debugging purposes. However,
 signatures over messages and assertions need a single canonical representation
 of the object to be signed as a bitstream. For RAINS, this is the CBOR
-representation canonicalized as in {{c14n}}; therefore alternate
-representations are always secondary to the CBOR data model. 
+representation canonicalized as in {{c14n}}; therefore alternate representations
+are always secondary to the CBOR data model. An alternate representation
+designed for textual manipulation of RAINS data is described in {{zonefiles}}.
```
Comment (britram) [inline draft-trammell-rains-protocol.md]: No it isn't; this was removed.

File: draft-trammell-rains-protocol.md [hunk 6]
```diff
@@ -688,18 +689,11 @@ the entire Message, generated as in {{signatures}}, and to be verified against
...
 the message.
 
-A Message map MAY contain a capabilities (1) key, for exposing capabilties of
-the message sender. The first Message sent from one peer to another MUST contain
+The value of the capabilities key, when present, is an array of Capabilities or
+the hash thereof. The first Message sent from one peer to another MUST contain
 the capabilities key. The capabilities mechanism is described in
 {{capabilities}}.
 
-A Message map MUST contain a token (2) key, whose value is a 16-byte array.
```
Comment (britram) [inline draft-trammell-rains-protocol.md]: Where did these go?

File: draft-trammell-rains-protocol.md [hunk 7]
```diff
@@ -735,37 +729,37 @@ The subject of an Assertion is identified by a name in three parts:
...
-  object of a given type, as well as the absence of such an association. Zones
-  are described in detail in {{zones}}. If signed, the assertions within a Zone
-  can also be treated as Singular Assertions; in this case they inherit zone and
-  context information from the containing zone.
-- Shards contain Singular Assertions for every object associated with every
+- Zones contain all Singular Assertions that have the same zone and context
+  values. The signature on a Zone can be used to prove both the existence of an
+  association between a subject name and an Object, as well as the absence of
+  such an association. Zones are described in detail in {{zones}}. If signed,
+  the Singular Assertions within a Zone can also be treated as Singular
```
Comment (britram) [inline draft-trammell-rains-protocol.md]:
```suggestion
  the Singular Assertions within a Zone can also be used on their own, as if they were contained within a Message directly; ```

File: draft-trammell-rains-protocol.md [hunk 8]
```diff
@@ -735,37 +729,37 @@ The subject of an Assertion is identified by a name in three parts:
...
+  Assertions; in this case they inherit zone and context information from the
```
Comment (britram) [inline draft-trammell-rains-protocol.md]:
```suggestion
 in this case they inherit zone and context information from the
```

File: draft-trammell-rains-protocol.md [hunk 9]
```diff
@@ -735,37 +729,37 @@ The subject of an Assertion is identified by a name in three parts:
...
+- Shards contain Singular Assertions for every Object associated with every
   subject name in a given lexicographic range of subject names within a given
   zone in a given context. The signature on a Shard can be used to prove the
-  nonexistance of an object of a given type for a name within its range. Shards
-  are described in detail in {{shards}}. If signed, the assertions within a
-  Shard can also be treated as Singular Assertions; in this case they inherit
-  zone and context information from the containing shard.
-- P-Shards (or Probabalistic Shards) contain a data structure that can be used
+  nonexistence of an Object for a subject name within its range. Shards are
+  described in detail in {{shards}}. If a Singular Assertions within a Shard is
```
Comment (britram) [inline draft-trammell-rains-protocol.md]:
```suggestion
  described in detail in {{shards}}. If a Singular Assertion within a Shard is
```

File: draft-trammell-rains-protocol.md [hunk 10]
```diff
@@ -911,56 +890,49 @@ The value of the context (6) key is a UTF-8 encoded string containing the name
...
 The value of the subject-zone (4) key is a UTF-8 encoded string containing the
-name of the zone in which the Assertions in the P-Shard is made and MUST end
-with '.' (the root zone).
+name of the zone in which the Singular Assertions in the P-Shard are made and
+MUST end with '.' (the root zone).
 
 The value of the context (6) key is a UTF-8 encoded string containing the name
-of the context in which the Assertions in the P-Shard are valid. Both the
-authority-part and the context-part MUST end with a '.'.
+of the context in which the Singular Assertions in the P-Shard are valid. Both
```
Comment (britram) [inline draft-trammell-rains-protocol.md]: The P-Shard doesn't contain Singular Assertions, it contains a Bloom filter against which names and objects can be checked. Rework this.

File: draft-trammell-rains-protocol.md [hunk 11]
```diff
@@ -979,33 +951,38 @@ are given in {{hash-functions}}.
...
+the subject context. The fourth element is the type code as in {{tabobj}} in
+{{obj-types}}. This encoded object is then hashed according to the specified
+hash algorithm. The hash algorithm's output is then split into nh equal length
+parts (2 for bloom-km-2, 4 for bloom-km-4, 8 for bloom-km-8), and these parts
+are used as indexes into the bitstring modulo the bitstring length. To add the
+Singular Assertion, all bits at the given indices are set to 1. To verify the
+Singular Assertion, all bits at the given indices are checked, and the Singular
+Assertion is taken to be in the filter if all bits are 1.
+
+[EDITOR's NOTE CFE: This is not how it is described in the km-paper
```
Comment (britram) [inline draft-trammell-rains-protocol.md]: to discuss

File: draft-trammell-rains-protocol.md [hunk 12]
```diff
@@ -979,33 +951,38 @@ are given in {{hash-functions}}.
...
+these Singular Assertions is considered to be the set of valid values at that
+point in time.
 
 ### Semantic of nonexistence proofs {#antiassertions}
 
 Shards, P-Shards and Zones can all be used to prove nonexistence during their
-validity. However, real naming systems are dynamic: an assertion might be
-created, altered, expired or revoked during the validity period of a shard,
-P-Shard or zone, leading to an inconsistency. Thus, a section proving
+validity. However, real naming systems are dynamic: a Singular Assertion might
```
Comment (britram) [inline draft-trammell-rains-protocol.md]: this is overcorrection: here, you actually should say Assertion.

```suggestion
validity. However, real naming systems are dynamic: an Assertion might
```

File: draft-trammell-rains-protocol.md [hunk 13]
```diff
@@ -911,56 +890,49 @@ The value of the context (6) key is a UTF-8 encoded string containing the name
...
 
 The value of the range (11) key, if present, is a two element array of strings
 or nulls (subject-name A, subject-name B). A MUST lexicographically sort before
 B. If A is null, the P-Shard begins at the beginning of the zone. If B is null,
-the shard ends at the end of the zone. The P-Shard MUST NOT be used to check
-the existence of any assertions whose subject names are equal to or sort before
-A, or are equal to or sort after B. If the range (11) key is not present, the
-P-shard covers then entire zone.
+the P-Shard ends at the end of the zone. The P-Shard MUST NOT be used to check
+the existence of any Singular Assertions whose subject names are equal to or
```
Comment (britram) [inline draft-trammell-rains-protocol.md]: overcorrection

```suggestion
the existence of Assertions about subject names equal to or
```

File: draft-trammell-rains-protocol.md [hunk 14]
```diff
@@ -911,56 +890,49 @@ The value of the context (6) key is a UTF-8 encoded string containing the name
...
 The value of the signatures (0) key is an array of one or more Signatures as
 defined in {{signatures}}. The signatures on the P-Shard are to be verified
-against the appropriate key for the Zone for which the P-Shard is cvalid in the
+against the appropriate key for the Zone for which the P-Shard is valid in the
 given context.
 
 The value of the subject-zone (4) key is a UTF-8 encoded string containing the
-name of the zone in which the Assertions in the P-Shard is made and MUST end
-with '.' (the root zone).
+name of the zone in which the Singular Assertions in the P-Shard are made and
```
Comment (britram) [inline draft-trammell-rains-protocol.md]: overcorrection

```suggestion
name of the zone within which the names represented in the P-Shard are contained, and
```

File: draft-trammell-rains-protocol.md [hunk 15]
```diff
@@ -979,33 +951,38 @@ are given in {{hash-functions}}.
...
-bloom-km-2, 4 for bloom-km-4, 8 for bloom-km-8), and these parts are used as
-indexes into the bitstring modulo the bitstring length. To add the assertion,
-all bits at the given indices are set to 1. To verify the assertion, all bits
-at the given indices are checked, and the assertion is taken to be in the
-filter if all bits are 1.
+The bloom-km-2, bloom-km-4 and bloom-km-8 datastructures generate a bitstring
+using a Bloom filter and the Kirsch-Mitzenmacher optimization
+{{BETTER-BLOOM-FILTER}}.
+
+To add or verify a Singular Assertion to a bloom-km structure, the Singular
```
Comment (britram) [inline draft-trammell-rains-protocol.md]: it's not a singular assertion, it's a name with an object.

File: draft-trammell-rains-protocol.md [hunk 16]
```diff
@@ -979,33 +951,38 @@ are given in {{hash-functions}}.
...
+Assertion is first encoded as a four-element CBOR array. The first element is
+the subject name. The second element is the subject zone. The third element is
+the subject context. The fourth element is the type code as in {{tabobj}} in
+{{obj-types}}. This encoded object is then hashed according to the specified
+hash algorithm. The hash algorithm's output is then split into nh equal length
+parts (2 for bloom-km-2, 4 for bloom-km-4, 8 for bloom-km-8), and these parts
+are used as indexes into the bitstring modulo the bitstring length. To add the
+Singular Assertion, all bits at the given indices are set to 1. To verify the
```
Comment (britram) [inline draft-trammell-rains-protocol.md]: see above.

File: draft-trammell-rains-protocol.md [hunk 17]
```diff
@@ -1247,12 +1222,12 @@ secured with PKIX {{!RFC5280}} certificates.
 |    3 | ee   | End-Entity Certificate           |
 
 A trust anchor certificate constraint specifies a certificate that MUST appear
-as the trust anchor for the certificate presented by the subject of the
-assertion on a connection attempt. An end-entity certificate constraint
-specifies a certificate that MUST be presented by the subject of the assertion
-on a connection attempt.
+as the trust anchor for the certificate presented by the subject of the Singular
```
Comment (britram) [inline draft-trammell-rains-protocol.md]: This is stylistic; it's not important to note the singularity of the assertion here. 

```suggestion
as the trust anchor for the certificate presented by the subject of the 
```

File: draft-trammell-rains-protocol.md [hunk 18]
```diff
@@ -1247,12 +1222,12 @@ secured with PKIX {{!RFC5280}} certificates.
...
+Assertion on a connection attempt. An end-entity certificate constraint
+specifies a certificate that MUST be presented by the subject of the Singular
```
Comment (britram) [inline draft-trammell-rains-protocol.md]: likewise
```suggestion
specifies a certificate that MUST be presented by the subject
```

File: draft-trammell-rains-protocol.md [hunk 19]
```diff
@@ -1247,12 +1222,12 @@ secured with PKIX {{!RFC5280}} certificates.
...
+Assertion on a connection attempt.
```
Comment (britram) [inline draft-trammell-rains-protocol.md]:
```suggestion
on a connection attempt.
```

File: draft-trammell-rains-protocol.md [hunk 20]
```diff
@@ -1328,56 +1303,60 @@ as in {{signatures}}. See {{public-key-management}} for more.
...
-- hashing assertions into Bloom filters for probabalistic shards (see {{p-shards}})
-- hashing assertion and message data as part of generating a MAC (see {{signatures}})
-- hashing assertion data for a confirmation query (see {{confirmation}})
-
-Where they are used in RAINS, hash functions are identified by a code given in
-{{tabhash}}. In this table, applicability "C" means the hash is valid for use
-in a certificate info object, "P" that it can be used for hashing assertions
-for P-shards, "S" that it can be used for hashing assertions and messages for
-signatures, and "Q" that it can be used for hashing assertions for confirmation queries.
+- hashing Singular Assertions into Bloom filters and check if a Singular
```
Comment (britram) [inline draft-trammell-rains-protocol.md]: singular assertions are not contained in p-shards, see above.

File: draft-trammell-rains-protocol.md [hunk 21]
```diff
@@ -1616,8 +1593,8 @@ element to be signed. The signing process is defined as follows:
 - Render the element to be signed into a canonical byte stream as specified in
   {{c14n}}.
 
-- Generate a signature on the resulting byte stream according to the algorithm
-  selected.
+- Generate a signature on the resulting byte stream according to the algorithm,
+  keyspace and key phase selected.
```
Comment (britram) [inline draft-trammell-rains-protocol.md]: signature is not dependent on keyspace and keyphase.

```suggestion
selected.
```

File: draft-trammell-rains-protocol.md [hunk 22]
```diff
@@ -1616,8 +1593,8 @@ element to be signed. The signing process is defined as follows:
 - Render the element to be signed into a canonical byte stream as specified in
   {{c14n}}.
 
-- Generate a signature on the resulting byte stream according to the algorithm
-  selected.
+- Generate a signature on the resulting byte stream according to the algorithm,
```
Comment (britram) [inline draft-trammell-rains-protocol.md]:
```suggestion
- Generate a signature on the resulting byte stream according to the algorithm
```

File: draft-trammell-rains-protocol.md [hunk 23]
```diff
@@ -1654,26 +1632,22 @@ To generate a canonicalized Singular Assertion:
 
 To generate a canonicalized Shard:
 
-- sort the objects array in each assertion contained in the assertions array as
-  for Singular Assertions, above.
+- sort the objects array in each Singular Assertion contained in the assertions
+  array as, above.
 - sort the assertions array by lexicographic order of the serialized
   canonicalized byte string representing the assertion. Note that this will
   cause the subject array to be sorted in lexicographic order of subject name,
```
Comment (britram) [inline draft-trammell-rains-protocol.md]:
```suggestion
  cause the assertions array to be sorted in lexicographic order of subject name,
```

File: draft-trammell-rains-protocol.md [hunk 24]
```diff
@@ -1654,26 +1632,22 @@ To generate a canonicalized Singular Assertion:
...
+- sort the objects array in each Singular Assertion contained in the assertions
+  array as, above.
 - sort the assertions array by lexicographic order of the serialized
   canonicalized byte string representing the assertion. Note that this will
-  cause the array to be sorted in lexicographic order of subject name,
-  as well.
-- sort the shards array by lexicographic order of the serialized canonicalized
-  byte string representing the shard. Note that this will
-  cause the array to be sorted in lexicographic order of shard range, as well.
+  cause the subject array to be sorted in lexicographic order of subject name,
```
Comment (britram) [inline draft-trammell-rains-protocol.md]:
```suggestion
  cause the assertions array to be sorted in lexicographic order of subject name,
```

File: draft-trammell-rains-protocol.md [hunk 25]
```diff
@@ -1682,8 +1656,9 @@ To generate a canonicalized P-Shard:
 
 To generate a canonicalized Message:
 
-- preserve the order of Sections within the Message
-- canonicalize each section as appropriate by following the canonicalization
+- preserve the order of the Sections (Assertions, Queries and Notifications)
```
Comment (britram) [inline draft-trammell-rains-protocol.md]: What is the value of this change? The intended meaning is that the order of Sections within a Message is significant; adding types of sections sounds, to me, like the intention is to order sections by type?

File: draft-trammell-rains-protocol.md [hunk 26]
```diff
@@ -1767,14 +1743,13 @@ algorithms, and so on.
...
 | urn:x-rains:tlssrv | Listens for TLS/TCP connections (see {{transport-tls}} |
-| urn:x-rains:udpsrv | Accepts messages via UDP (see {{transport-udp}})       |
 
 A RAINS server MUST NOT assume that a peer server supports a given capability
 unless it has received a message containing that capability from that server. An
 exception are the capabilities indicating that a server listens for connections
 using a given transport protocol; servers and clients can also learn this
-information from RAINS itself (given redirection and service-info assertions for
-a named zone) or from external configuration.
+information from RAINS itself (given redirection and service-info Singular
```
Comment (britram) [inline draft-trammell-rains-protocol.md]: stylistic: singularity is unimportant here.

```suggestion
information from RAINS itself (given redirection and service-info 
```

File: draft-trammell-rains-protocol.md [hunk 27]
```diff
@@ -1791,20 +1766,12 @@ send it.
...
-over a ubiquitously-available and well-understood security and transport layer.
-
-RAINS over UDP allows RAINS messages to be exchanged with the same statelessness
-and confidentiality properties as DNS over UDP, reducing state requirements at
-RAINS clients and servers for persistent TLS connection, but without
-confidentiality protection, and with limits on the size of messages which can be
-transferred. RAINS servers MUST support TLS over TCP; other transports can be
-negotiated using the capabilities mechanism (see {{capabilities}}) after
-bootstrapping using TLS 1.3 (or later).
+This document so far defines one transport binding for RAINS: TLS-over-TCP in
```
Comment (britram) [inline draft-trammell-rains-protocol.md]:
```suggestion
This document defines one transport binding for RAINS: TLS-over-TCP in
```

File: draft-trammell-rains-protocol.md (hunk 2 above)
Comment (fehlmach) [reply draft-trammell-rains-protocol.md]: It is described after all Sections are introduced. Should we move it up

File: draft-trammell-rains-protocol.md (hunk 6 above)
Comment (fehlmach) [reply draft-trammell-rains-protocol.md]: I moved it up where the other Message elements are mentioned

File: draft-trammell-rains-protocol.md (hunk 25 above)
Comment (fehlmach) [reply draft-trammell-rains-protocol.md]: We never state what a 'Section' is. Better define it properly when we introduce Messages and remove my additions here

File: draft-trammell-rains-protocol.md (hunk 1 above)
Comment (fehlmach) [reply draft-trammell-rains-protocol.md]: rolled back

File: draft-trammell-rains-protocol.md (hunk 10 above)
Comment (fehlmach) [reply draft-trammell-rains-protocol.md]: True, my bad...

**Pull request merged**